docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+* prediction must not change the state of the cluster
+* provide general framework usable for any scheduling algorithm
+
+Scheduling process uses two sets of functions: predicates and priority functions.
+Predicates select all nodes suitable for scheduling.
+Priority functions choose the most suitable node from the selected nodes.
+Both predicates and priority functions may need to query the current state of the cluster during execution.
+Queries can include node's free cpu and memory, volume claims, annotations, etc.
+Once a pod gets scheduled on a node, it can change its free resources and change the decision basis of the next scheduling iteration.
+
+Thus, when designing the predictor, one has to take into account the current cluster state and compute its next without changing the current one.
+I.e. simulate the pod scheduling and binding. The simulation consists of (not exclusively in this order):
+
+* estimation of the current cluster state (node info, available volume claims, etc.)
+* execution of one scheduling iteration
+* update of scheduler caches based on scheduled pod
+
+Some consumed resources do not have to be specified in the pod (e.g. volume claims).
+These kind of consumed resources must be simulated accordingaly.
+
+Once the current state is captured, prediction is no longer dependent on the cluster state.
+Thus, the estimation does not have to correspond to the real scheduling since:
+
+* some nodes can be deleted, some nodes can be evacuated, some nodes can be re-labeled/re-annotated
+* some pods can get evicted in the time of prediction
+* remaining capacity of nodes can increase/decrease
+
+### Other aspects to consider
+
+One could take into account shared resources [3] when predicting free capacity (with fragmenation) as well.
+Currently, limit and requests are supported for cpu and memory resource types only.
+In future, the types can be extended to node disk space resources and custom resource types [4].
+
+Another thing to keep in mind is overcommitment with multiple levels of QoS.
+Or consider compressible vs. incompressible resources.
+
+#### Cluster resource subdivision
+
+Allow cluster resources to be subdivided (https://github.com/kubernetes/kubernetes/issues/442).
+
+#### Resource quota
+
+Resource quota [5] can influence number of pods that can be scheduled per namespace [6] as well.
+Though, this limitation is not hardware specific, it can be taken into account in future implementations.
+
+Example: In a cluster with a capacity of 32 GiB RAM, and 16 cores, let team A use 20 Gib and 10 cores, let B use 10GiB and 4 cores, and hold 2GiB and 2 cores in reserve for future allocation.
+
+#### Federated scheduling
+Out of scope of the document.
+Can be implemented as an agreggation of cluster capacities of individual subclusters.
+
+## Implemenation
+
+Goals:
+
+* Build the implementation over the existing code.
+* Run the final implementation as a self-standing application in a pod (under `kube-system` namespace).
```

Comment:
It reminds me multiple schedulers design. Maybe, provide each scheduler as a separate projects. All schedulers than get initialize as a part of the predictor framework. The local client will then emulate the Apiserver + Kubelet. Or it can be more general and emulate any other environment that corresponds to prediction requirements.